docs/proposals/thin-volumes.md

```diff
@@ -0,0 +1,189 @@
++   string thin_pool = 4;
++   uint64 overprovision_bytes = 5; // Size up and above thin pool capacity
+}
+```
+- Resize and remove of Logical Volumes will use the same API as the underlying calls on volume group or thinpool only needs Logical Volume ID.
+
+
+#### Monitoring thin-pools
+- We need to constantly monitor the thin-pools and ensure that it does not run out of space.
+- `/etc/lvm/lvm.conf` settings can allow auto growth of the thin pool when required. By default, the threshold is 100% which means that the pool will not grow.
+- If we set this to, 75%, the Thin Pool will autoextend when the pool is 75% full. It will increase by the default percentage of 20% if the value is not changed. We can see these settings using the command grep against the file.
+
+```go=
+# grep -E ‘^\s*thin_pool_auto’ /etc/lvm/lvm.conf
+ thin_pool_autoextend_threshold = 100
+ thin_pool_autoextend_percent = 20
+```
+- We assume that user has the threshold of 100% with no autogrow.
+- Topolvm will regularly monitor the used thin-pool space for `pool data` and `metadata`.
+- These usages will be exported as metrics so that admin/opeartor can analyse them and take precautionary steps whens usage exeeds certain threshold limit.
+- Command to be used for montoring:
+``` go=
+   lvs
+   LV          VG   Attr       LSize   Pool  Origin Data%  Meta%  Move Log Cpy%Sync Convert
+   lv1         vg1  -wi-ao---- 600.00m
+   thin_volume vg1  Vwi-a-tz--   1.00g tpool        3.20
+   tpool       vg1  twi-aotz-- 200.00m              16.38  1.37
+```
+- Topolvm will only monitor the thin-pools provided in the lvmd.yaml config file.
+
+#### Capacity awareness
+- Topolvm uses capacity awareness to ensure that new pod is created on the node where available capacity of the volume group is larger than the requested capacity.`capacity.topolvm.cybozu.com/<device-class>` annotation is added on each node which consists of the available capacity on the volume group.
+- Creation of thin logical volumes doesn't update the available device class capacity until the user actually uses storage on the thin volumes.
+- So we need to ensure that all the thin logical volumes don't get created on the same node.
+- This can be done by using the `overprovisionRatio`.
+- Value of `capacity.topolvm.cybozu.com/<device-class>_mlog<thin-pool>` annotation will be calculated as:
+  `(thin-pool-size * overprovisionedRatio)-(sum-of-all-thin-lv-sizes-in-thin-pool)`
+- Available thin pools on the volume groups will be obtained from lvmd.yaml config file.
+
+
+
+## Open Questions
+- Should topolvm create the thin-pools instead of the admin/operator?
+
+- If multiple thin-pools are supported for each volume group, then how to decide which thin-pool should be used to create the thin logical volumes.
+    - User should decide which thin-pool to use. This can done via a StorageClass parameter.
+
+- How to do capacity awareness of thin-pools when Volume groups have both thick and thin logical volumes.
```

Comment:
> The capacity aware scheduling code will be complicated very much if there are normal pools and thin pools at the same time.

Agree that presence of both thin and thick pools will complicate the scheduling decisions. 

> In addition, it's difficult to create practical logic in such case. So, how about adding a description to prohibit the co-existence of normal pool and thin pool in docs/limitations.md?

Do you mean that entire topolvm cluster can only be created either using thin-pools or thick pool and not with a combination of both?   If we do this, then existing topolvm customers won't be able to make use of the thin volumes feature. This will be a big disadvantage.

Reply:
@cupnes
- the idea here is to have distinct annotations for thin-pool and remaining space of VG (current method) which can be used to take scheduling decisions based on the ask for thin/thick logical volume

Reply:
@sp98 
> Do you mean that entire topolvm cluster can only be created either using either thin-pools or thick pools and not with a combination of both?

No, I don't. I mean that one volume group is for either thin volume or thick volume. VGs for thin volumes and VGs for thick pools can be in same TopoLVM cluster.

@leelavg 
I understood.

Reply:
> No, I don't. I mean that one volume group is for either thin volume or thick volume. VGs for thin volumes and VGs for thick pools can be in same TopoLVM cluster.

Thanks for the clarification @cupnes . I see you have acknowledged one of the [comments](https://github.com/topolvm/topolvm/pull/442#discussion_r822538501) provide by my colleague @leelavg about how we can use both thin and thick volumes on the same Volume group. 

Do you see any complexities for having both thick and thin pools on the same VG? We would like to have this because it will help existing customers using thick volumes to create thin-pools and create thin-volumes on them.

Reply:
@sp98 Thank you for additional information. I understood the need. Sorry, I will reply on this discussion next week.

Reply:
@cupnes - in particular we would like to keep existing thick-LVs on the VG, then add a thin-pool and from that point onwards only add thin-LVs. Thus the scheduling algorithm wouldn't need to consider this VG any more for new thick-LVs.

Reply:
Is there really a need for thin and thick to coexist in the same VG, or is it more about trying to help clusters upgrade and starting using thin after they had been created with thick? The thin/thick modes are so different, my first impression is that VGs should really be dedicated to either thick or thin. Instead of trying to support both in the same VG, perhaps there is a simpler migration path to convert thick VGs to thin VGs, or truncate unused space in the thick VG to create a new thin VG.

Reply:
> ... or is it more about trying to help clusters upgrade and starting using thin after they had been created with thick? 

That's the only use case for us AFAIK.

Reply:
The resolution is to have multiple device classes per volume group and each device class is mapped to a single thin pool on that volume group.  Marking this conversation as resolved.